File: Leetcode/Python3/771. Jewels and Stones/solution.py
# https://leetcode.com/problems/jewels-and-stones/

from collections import defaultdict

# Defaultdict
class Solution: 
    def numJewelsInStones(self, jewels: str, stones: str) -> int:
        freq = defaultdict(int)
        count = 0

        for stone in stones:
            freq[stone] += 1
        
        for jewel in jewels:
            count += freq[jewel]
        return count

# Set
class Solution: 
    def numJewelsInStones(self, jewels: str, stones: str) -> int:
        jewel_set = set(jewels)
        count = 0
        for stone in stones:
            if stone in jewel_set:
                count += 1
        return count 

# Comprehension
class Solution: 
    def numJewelsInStones(self, jewels: str, stones: str) -> int:
        return sum(s in jewels for s in stones)

File: Leetcode/Typescript/771. Jewels and Stones/solution.ts
// https://leetcode.com/problems/jewels-and-stones/description/

function numJewelsInStones(jewels: string, stones: string): number {
  const jewelSet = new Set<string>(jewels);
  let count = 0;

  for (const stone of stones) {
    if (jewelSet.has(stone)) {
      count += 1;
    }
  }

  return count;
}
